src/java/org/apache/cassandra/service/accord/AccordCommandStore.java

```diff
@@ -248,10 +407,113 @@ public void completeOperation(AccordSafeCommandStore store,
+    {
+        switch (keysOrRanges.domain())
+        {
+            case Key:
+            {
+                AbstractKeys<Key, ?> keys = (AbstractKeys<Key, ?>) keysOrRanges;
+                for (Key key : keys)
+                {
+                    if (!slice.contains(key)) continue;
+
+                    accumulate = map.apply(fromRangeSummary(key, rangesToCommands.search(key)), accumulate);
+                    if (accumulate.equals(terminalValue))
+                        return accumulate;
+                }
+            }
+            break;
+            case Range:
+            {
+                AbstractRanges<?> ranges = (AbstractRanges<?>) keysOrRanges;
+                ranges = ranges.slice(slice, Routables.Slice.Minimal);
+                for (Range range : ranges)
+                {
+                    accumulate = map.apply(fromRangeSummary(range, rangesToCommands.search(Interval.create(range.start(), range.end()))), accumulate);
+                    if (accumulate.equals(terminalValue))
+                        return accumulate;
+                }
+            }
+            break;
+            default:
+                throw new AssertionError("Unknown domain: " + keysOrRanges.domain());
+        }
+        return accumulate;
+    }
+
+    private void maybeUpdateRangeIndex(Map<TxnId, AccordSafeCommand> commands)
+    {
+        for (Map.Entry<TxnId, AccordSafeCommand> e : commands.entrySet())
+        {
+            TxnId txnId = e.getKey();
+            if (txnId.domain() != Routable.Domain.Range)
+                continue;
+            Command current = e.getValue().current();
+            if (current.saveStatus() == SaveStatus.NotWitnessed)
+                continue; // don't know the range/dependencies, so can't cache
+            PartialTxn txn = current.partialTxn();
+            Seekables<?, ?> keys = txn.keys();
+            if (keys.domain() != Routable.Domain.Range)
+                throw new AssertionError("Found a Range Transaction that had non-Range keys: " + current);
+            if (keys.isEmpty())
+                throw new AssertionError("Found a Range Transaction that has empty keys: " + current);
+            PartialDeps deps = current.partialDeps();
+            List<TxnId> dependsOn = deps == null ? Collections.emptyList() : deps.txnIds();
+
+            RangeCommandSummary summary = new RangeCommandSummary(txnId, current.saveStatus(), current.executeAt(), dependsOn);
+            Ranges ranges = (Ranges) keys;
+            put(ranges, summary);
+        }
+    }
+
+    private void put(Ranges ranges, RangeCommandSummary summary)
+    {
+        IntervalTree.Builder<RoutableKey, RangeCommandSummary, Interval<RoutableKey, RangeCommandSummary>> builder = rangesToCommands.unbuild();
+        //TODO double check this tree has same inclusive/exclusive semantics as this range...
```

Comment:
this looks inclusive on both sides, and our range tend to be `(, ]`, so not a perfect match... will look to improve